patches/api/0101-Add-support-for-multiple-plugin-folders.patch

```diff
@@ -0,0 +1,295 @@
+ 
++    /**
++     * Loads the plugins contained within the specified directories
++     *
++     * @param directories Directories to check for plugins
++     * @return A list of all plugins loaded
++     */
++    public Plugin[] loadPlugins(List<File> directories);
++
+     /**
+      * Disables all the loaded plugins
+      */
+diff --git a/src/main/java/org/bukkit/plugin/SimplePluginManager.java b/src/main/java/org/bukkit/plugin/SimplePluginManager.java
+index 5e58e8f8..d89e003b 100644
+--- a/src/main/java/org/bukkit/plugin/SimplePluginManager.java
++++ b/src/main/java/org/bukkit/plugin/SimplePluginManager.java
+@@ -3,22 +3,13 @@ package org.bukkit.plugin;
+ import java.io.File;
+ import java.lang.reflect.Constructor;
+ import java.lang.reflect.Method;
+-import java.util.ArrayList;
+-import java.util.Collection;
+-import java.util.HashMap;
+-import java.util.HashSet;
+-import java.util.Iterator;
+-import java.util.LinkedHashMap;
+-import java.util.LinkedList;
+-import java.util.List;
+-import java.util.Map;
+-import java.util.Set;
+-import java.util.WeakHashMap;
++import java.util.*;
+ import java.util.logging.Level;
+ import java.util.regex.Matcher;
+ import java.util.regex.Pattern;
+ 
+ import org.apache.commons.lang.Validate;
++import org.bukkit.Bukkit;
+ import org.bukkit.Server;
+ import org.bukkit.command.Command;
+ import org.bukkit.command.PluginCommandYamlParser;
+@@ -47,7 +38,6 @@ public final class SimplePluginManager implements PluginManager {
+     private final Map<Pattern, PluginLoader> fileAssociations = new HashMap<Pattern, PluginLoader>();
+     private final List<Plugin> plugins = new ArrayList<Plugin>();
+     private final Map<String, Plugin> lookupNames = new HashMap<String, Plugin>();
+-    private static File updateDirectory = null;
+     private final SimpleCommandMap commandMap;
+     private final Map<String, Permission> permissions = new HashMap<String, Permission>();
+     private final Map<Boolean, Set<Permission>> defaultPerms = new LinkedHashMap<Boolean, Set<Permission>>();
+@@ -105,91 +95,109 @@ public final class SimplePluginManager implements PluginManager {
+      * @param directory Directory to check for plugins
+      * @return A list of all plugins loaded
+      */
++    @Override
+     public Plugin[] loadPlugins(File directory) {
+-        Validate.notNull(directory, "Directory cannot be null");
+-        Validate.isTrue(directory.isDirectory(), "Directory must be a directory");
++        return loadPlugins(Collections.singletonList(directory));
++    }
++
++    /**
++     * Loads the plugins contained within the specified directories
++     *
++     * @param directories Directories to check for plugins
++     * @return A list of all plugins loaded
++     */
++    public Plugin[] loadPlugins(List<File> directories) {
++        Validate.notNull(directories, "Directories cannot be null");
++        if (directories.isEmpty()) {
++            return new Plugin[0];
```

Comment:
Previous logic threw an error for missing directories, need to preserve this.

Reply:
What do you mean?